README.md

# Fast_Zarinpal

Quick implementation of Zarin Pal payments with Fast_ZarinPal


![Logo](https://s5.uupload.ir/files/seraaj/Flutter_Downloads/FlutterPlus.ir/fast_ZarinPal.png)

# 🔧 Installation

   
## 1.Add Fast Zarinpal  to your project
```bash
flutter pub add fast_zarinpal
```
## 2.Add your callback in AndroidManifest.xml 
Add the following code in the Android Manifest file.  `android/app/main/AndroidManifest.xml` 
```xml
<intent-filter>
<action android:name="android.intent.action.VIEW" />
<category android:name="android.intent.category.DEFAULT" />
<category android:name="android.intent.category.BROWSABLE" />
<data 
android:scheme="payment"
android:host='yourdomain.com' />
</intent-filter>
```
❓You can enter your desired address instead of `yourdomain.com`.
If this address is not registered for another application, it will not be a problem.
For reference, the example app's AndroidManifest.xml file can be found [here](https://github.com/3raaj/fast_zarinpal/blob/47cee3d637fb3e996d2eedcdf8ec79e5da898937/example/android/app/src/main/AndroidManifest.xml#L23)
## Example

```dart
 FastZarinPal(
          callback: 'yourdomain.com',
          merchantID: 'e8188181-8181-8181-8181-818181818181',
          paymentTitle: 'خرید اشتراک ماهانه ',
          priceToPay: 125000,
          totalPurchaseAmount: 150000,
          dicount: 25000,
          onPaymentFailedWidgets: (amount) =>
              FailedPaymentWidget(themeData: themeData),
          onSuccessWidgets: (amount, refID) {
            return const SuccessPaymentWidget();
          },
),
class SuccessPaymentWidget extends StatelessWidget {
  const SuccessPaymentWidget({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        const Icon(
          Icons.check,
          color: Colors.green,
        ),
        const Text('پرداخت شما موفق بود ! '),
        const SizedBox(
          height: 12,
        ),
        const Divider(),
        TextButton(
            onPressed: () {}, child: const Text('بازگشت به صفحه فروشگاه')),
      ],
    );
  }
}

class FailedPaymentWidget extends StatelessWidget {
  const FailedPaymentWidget({
    Key? key,
    required this.themeData,
  }) : super(key: key);

  final ThemeData themeData;

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Icon(
          Icons.error,
          color: themeData.errorColor,
        ),
        Text(
          'پرداخت شما ناموفق بود :(',
          style: themeData.textTheme.subtitle1!.copyWith(
            color: themeData.errorColor,
          ),
        ),
        const Divider(),
        TextButton(onPressed: () {}, child: const Text('بازگشت و تلاش مجدد')),
      ],
    );
  }
}       
```


## Feedback

If you have any feedback, please reach out to us at info@flutterplus.ir




CHANGELOG.md
## 0.1.2

* update README.md +2 



test/fast_zarinpal_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:fast_zarinpal/fast_zarinpal.dart';

void main() {
  test('adds one to input values', () {
    final fastZarinPal = FastZarinPal(
      callback: 'yourdomain.com',
      merchantID: 'testMID',
      onPaymentFailedWidgets: (amount) {
        return const Text('test -> payment failed');
      },
      onSuccessWidgets: (amount, refID) {
        return const Text('test -> success payment');
      },
      paymentTitle: 'payment title',
      priceToPay: 1,
      totalPurchaseAmount: 2,
      dicount: 0,
    );

    expect(fastZarinPal.totalPurchaseAmount, 2);
    expect(fastZarinPal.callback, 'yourdomain.com');
    expect(
        fastZarinPal.onSuccessWidgets, const Text('test -> success payment'));
    expect(fastZarinPal.priceToPay, 1);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:zarrinpal_payment_test/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/fast_zarinpal.dart
library fast_zarinpal;

/// A Calculator.
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:persian_number_utility/persian_number_utility.dart';
import 'package:uni_links/uni_links.dart';
import 'package:url_launcher/url_launcher.dart';
import 'package:zarinpal/zarinpal.dart';

import 'bloc/payment_bloc.dart';

class FastZarinPal extends StatefulWidget {
  ///The amount that will actually be paid
  final double priceToPay;
  //Total Amount
  final double totalPurchaseAmount;
  ///If the discount is empty, it will be considered zero. In any case, if there is a discount, you must subtract it from "priceToPay".
  final double? dicount;
  ///The callback you defined in the AndroidManifest.xml file. It can be any desired value.
  final String callback;
  ///In case of successful payment, what widgets should be displayed?
  final Widget Function(String amount, String refID) onSuccessWidgets;
  ///What widgets should be displayed in case of unsuccessful payment?
  final Widget Function(String amount) onPaymentFailedWidgets;
  ///Merchant ID You can save in a separate const.dart file.
  final String merchantID;
  ///The title displayed on the online checkout page.
  final String paymentTitle;
  /// A title displayed on the payment result page.  default value is 'نتیجه پرداخت',
  final  String paymentResultTitle;
   const FastZarinPal(
      {super.key,
      this.paymentResultTitle = 'نتیجه پرداخت', 
      required this.priceToPay,
      required this.paymentTitle,
      required this.merchantID,
      this.dicount,
      required this.totalPurchaseAmount,
      required this.onSuccessWidgets,
      required this.onPaymentFailedWidgets,
      required this.callback});

  @override
  State<FastZarinPal> createState() => _FastZarinPalState();
}

// final PaymentBloc paymentBloc = PaymentBloc(selectedSubMode);
class _FastZarinPalState extends State<FastZarinPal> {
  final TextEditingController discountCodeTextEditing = TextEditingController();
  PaymentBloc paymentBloc = PaymentBloc();

  late PaymentRequest paymentRequest;

  bool isLoading = false;
  bool isLoadingPay = false;
  Uri? payUri;
  @override
  void dispose() {
    discountCodeTextEditing.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    initPlatformStateForUriUniLinks();
    final ThemeData themeData = Theme.of(context);
    return Scaffold(
      body: Directionality(
        textDirection: TextDirection.rtl,
        child: BlocProvider(
          create: (context) {
            paymentBloc.add(PaymentStarted());
            paymentBloc.stream.forEach((state) {
              if (state is CloseFromPayment) {
                Navigator.of(context).pop();
              }
            });
            return paymentBloc;
          },
          child: BlocBuilder<PaymentBloc, PaymentState>(
            builder: (context, state) {
              if (state is ShowPaymentResultScreen ||
                  state is SuccessAddPaymentHistory ||
                  state is ErrorAddPaymentHistory) {
                return SingleChildScrollView(
                  child: Center(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.center,
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        const SizedBox(
                          height: 100,
                        ),
                        Text(
                          'نتیجه پرداخت',
                          style: themeData.textTheme.subtitle1,
                        ),
                        const SizedBox(
                          height: 26,
                        ),
                        state is ShowPaymentResultScreen
                            ? state.isPaymentSuccess
                                ? widget.onSuccessWidgets(
                                    state.amount.toString(), state.refID!)
                                : !state.isPaymentSuccess
                                    ? widget.onPaymentFailedWidgets(
                                        state.amount.toString())
                                    : const SizedBox()
                            : const SizedBox(),
                        const SizedBox(
                          height: 20,
                        ),
                      ],
                    ),
                  ),
                );
              } else if (state is LoadingPayment || state is CloseFromPayment) {
                if (state is CloseFromPayment) {
                  debugPrint('Close From Payment ! :/ ');
                }
                return const Center(child: AnimtedLoading());
              } else {
                return Center(
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      Container(
                        padding: const EdgeInsets.all(12),
                        margin: const EdgeInsets.all(12),
                        width: MediaQuery.of(context).size.width,
                        height: MediaQuery.of(context).size.height * 0.5,
                        decoration: BoxDecoration(
                            borderRadius: BorderRadius.circular(12),
                            color: Colors.white,
                            boxShadow: const [
                              BoxShadow(
                                color: Colors.black12,
                                blurRadius: 12,
                              ),
                            ]),
                        child: Column(
                          children: [
                            Row(
                                mainAxisAlignment:
                                    MainAxisAlignment.spaceBetween,
                                children: [
                                  const Text('جزئیات خرید'),
                                  GestureDetector(
                                    onTap: () {
                                      Navigator.of(context).pop();
                                    },
                                    child: Container(
                                        padding: const EdgeInsets.all(4),
                                        decoration: BoxDecoration(
                                            color: ThemeColors.errorColor,
                                            borderRadius:
                                                BorderRadius.circular(10)),
                                        child: const Icon(
                                          Icons.close,
                                          color: Colors.white,
                                          size: 12,
                                        )),
                                  ),
                                ]),
                            Row(
                              children: [
                                Icon(
                                  Icons.sell,
                                  color: themeData.primaryColor,
                                ),
                                const SizedBox(
                                  width: 4,
                                ),
                                Text(widget.paymentTitle),
                              ],
                            ),
                            const Divider(),
                            const SizedBox(
                              height: 36,
                            ),
                            Row(
                              children: [
                                Flexible(
                                  flex: 1,
                                  child: Column(
                                    children: [
                                      Row(
                                        children: [
                                          const Text('مبلغ  کل خرید'),
                                          Expanded(
                                              child: Row(
                                            mainAxisAlignment:
                                                MainAxisAlignment.end,
                                            children: [
                                              Text(
                                                '${widget.totalPurchaseAmount.toInt()} تومان',
                                                style: themeData
                                                    .textTheme.caption!
                                                    .copyWith(
                                                  color:
                                                      const Color(0xff72A3E4),
                                                ),
                                              ),
                                            ],
                                          )),
                                        ],
                                      ),
                                      const Divider(
                                        height: 16,
                                        thickness: 1,
                                      ),
                                      Row(
                                        children: [
                                          const Text('تخفیف'),
                                          Expanded(
                                            child: Row(
                                              mainAxisAlignment:
                                                  MainAxisAlignment.end,
                                              children: [
                                                Text(
                                                  '${widget.dicount != null ? widget.dicount!.toInt() : 0} تومان',
                                                  style: themeData
                                                      .textTheme.caption!
                                                      .copyWith(
                                                    color:
                                                        ThemeColors.errorColor,
                                                  ),
                                                ),
                                              ],
                                            ),
                                          ),
                                        ],
                                      )
                                    ],
                                  ),
                                ),
                                Flexible(
                                  flex: 1,
                                  child: Column(
                                    children: [
                                      Row(
                                        mainAxisAlignment:
                                            MainAxisAlignment.center,
                                        children: [
                                          Text(
                                            'مبلغ پرداختی ',
                                            style: themeData
                                                .textTheme.subtitle1!
                                                .copyWith(
                                              color: const Color(0xff3ECB6E),
                                            ),
                                          ),
                                        ],
                                      ),
                                      const SizedBox(
                                        height: 12,
                                      ),
                                      Row(
                                        mainAxisAlignment:
                                            MainAxisAlignment.center,
                                        children: [
                                          Flexible(
                                            child: Text(
                                              widget.priceToPay
                                                  .toInt()
                                                  .toString()
                                                  .seRagham(),
                                              style: themeData
                                                  .textTheme.headline6!
                                                  .copyWith(
                                                      fontWeight:
                                                          FontWeight.w900),
                                            ),
                                          ),
                                          const SizedBox(
                                            width: 4,
                                          ),
                                          const Text('تومان')
                                        ],
                                      ),
                                    ],
                                  ),
                                ),
                              ],
                            ),
                            const Expanded(
                              child: SizedBox(),
                            ),
                            Row(
                              children: [
                                Expanded(
                                  child: ElevatedButton(
                                    style: ButtonStyle(
                                        backgroundColor:
                                            MaterialStateProperty.all(
                                                const Color(0xff34CC73))),
                                    onPressed: () async {
                                      setState(() {
                                        paymentRequest = PaymentRequest()
                                          ..setIsSandBox(
                                              true) // if your application is in developer mode, then set the sandBox as True otherwise set sandBox as false
                                          ..setMerchantID(widget.merchantID)
                                          ..setCallbackURL(
                                              "payment://${widget.callback}");
                                        //The callback can be an android scheme or a website URL, you and can pass any data with The callback for both scheme and  URL
                                      });

                                      paymentRequest
                                          .setAmount(widget.priceToPay);

                                      paymentRequest
                                          .setDescription(widget.paymentTitle);
                                      // Call Start payment
                                      setState(() {
                                        isLoadingPay = true;
                                      });
                                      ZarinPal().startPayment(
                                        paymentRequest,
                                        (status, paymentGatewayUri) async {
                                          if (status == 100) {
                                            launchUrl(
                                              Uri.parse(paymentGatewayUri!),
                                              mode: LaunchMode
                                                  .externalNonBrowserApplication,
                                            );
                                          }
                                        },
                                      );
                                    },
                                    child: isLoadingPay
                                        ? const SizedBox(
                                            height: 20,
                                            width: 20,
                                            child: Center(
                                              child: CircularProgressIndicator(
                                                color: Colors.white,
                                                strokeWidth: 5,
                                              ),
                                            ),
                                          )
                                        : const Text('پرداخت آنلاین'),
                                  ),
                                ),
                              ],
                            ),
                          ],
                        ),
                      ),
                    ],
                  ),
                );
              }
            },
          ),
        ),
      ),
    );
  }

  initPlatformStateForUriUniLinks() async {
    uriLinkStream.listen((Uri? uri) {
      var urlQuery = uri!.queryParametersAll.entries.toList();
      if (urlQuery[1].value[0] == "OK") {
        ZarinPal().verificationPayment(
            urlQuery[1].value[0], urlQuery[0].value[0], paymentRequest,
            (isPaymentSuccess, refID, paymentRequest) {
          if (isPaymentSuccess) {
            paymentBloc.add(
              NavigateFromPaymentClicked(
                  paymentRequest: paymentRequest,
                  isPaymentSuccess: isPaymentSuccess,
                  refID: refID),
            );
          }
        });
      } else if (urlQuery[1].value[0] == "NOK") {
        paymentBloc.add(
          NavigateFromPaymentClicked(
              paymentRequest: paymentRequest,
              isPaymentSuccess: false,
              refID: 'Undefined!'),
        );
      }
    }, onError: (error) {
      debugPrint('Error ==========================');
    });
  }
}

class ThemeColors {
  static const Color secoundryColor = Colors.white;
  static const Color primaryColor = Color(0xffFCBE11);

  // ignore: use_full_hex_values_for_flutter_colors
  static const Color primayTextColor = Color(0xff4210752);
  static const Color errorColor = Color(0xffBD272D);
  static const Color profileIconsColor = Color(0xffAFAFAF);
  static const Color successColor = Color(0xff1A8755);
}

class AnimtedLoading extends StatelessWidget {
  const AnimtedLoading({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const Center(
      child: CircularProgressIndicator(
        strokeWidth: 4,
        color: Colors.green,
      ),
    );
  }
}


lib/bloc/payment_bloc.dart
import 'package:equatable/equatable.dart';
import 'package:flutter_bloc/flutter_bloc.dart';

import 'package:zarinpal/zarinpal.dart';

part 'payment_event.dart';
part 'payment_state.dart';

class PaymentBloc extends Bloc<PaymentEvent, PaymentState> {
  PaymentBloc() : super(PaymentInitial()) {
    on<PaymentEvent>((event, emit) async {
      if (event is PaymentStarted) {
        emit(ShowCartScreen());
      } else if (event is PaymentClosed) {
        emit(CloseFromPayment());
      } else if (event is NavigateFromPaymentClicked) {
        emit(LoadingPayment());
        final bool isPaymentSuccess = event.isPaymentSuccess;
        final String description = event.paymentRequest.description!;
        final num amount = event.paymentRequest.amount!;

        emit(ShowPaymentResultScreen(
          amount: amount,
          refID: event.refID,
          description: description,
          isPaymentSuccess: isPaymentSuccess,
        ));
      }
    });
  }
}


lib/bloc/payment_state.dart
part of 'payment_bloc.dart';

abstract class PaymentState extends Equatable {
  const PaymentState();

  @override
  List<Object> get props => [];
}

class PaymentInitial extends PaymentState {}

class ShowPaymentResultScreen extends PaymentState {
  final bool isPaymentSuccess;
  final String? refID;
  final String description;
  final num amount;

  const ShowPaymentResultScreen(
      {
        required this.refID, 
        required this.isPaymentSuccess,
      required this.description,
      required this.amount});
  @override
  List<Object> get props => [isPaymentSuccess, description, amount];
}

class SuccessAddPaymentHistory extends PaymentState {
  final bool isPaymentSuccessAdded;
  final bool isPaymentSuccess;
  final String description;
  final num amount;
  const SuccessAddPaymentHistory(
      {required this.isPaymentSuccess,
      required this.description,
      required this.amount,
      required this.isPaymentSuccessAdded});
  @override
  List<Object> get props =>
      [isPaymentSuccess, isPaymentSuccessAdded, description];
}

class ErrorAddPaymentHistory extends PaymentState {
  final bool isPaymentSuccessAdded;
  final bool isPaymentSuccess;
  final String description;
  final num amount;
  const ErrorAddPaymentHistory(
      {required this.isPaymentSuccess,
      required this.description,
      required this.amount,
      required this.isPaymentSuccessAdded});
  @override
  List<Object> get props =>
      [isPaymentSuccessAdded, isPaymentSuccess, description];
}

class LoadingPayment extends PaymentState {}

class ShowCartScreen extends PaymentState {}

class CloseFromPayment extends PaymentState{}

example/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "zarrinpal_payment_test");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "zarrinpal_payment_test");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"zarrinpal_payment_test", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/main.dart
import 'package:fast_zarinpal/fast_zarinpal.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    final ThemeData themeData = Theme.of(context);
    return MaterialApp(
      title: 'Fast Zarinpal Test',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: FastZarinPal(
          callback: 'yourdomain.com',
          merchantID: 'e8181811-8181-8181-bc31-813fe9818181',
          paymentTitle: 'خرید اشتراک ماهانه ',
          priceToPay: 125000,
          totalPurchaseAmount: 150000,
          dicount: 25000,
          onPaymentFailedWidgets: (amount) =>
              FailedPaymentWidget(themeData: themeData),
          onSuccessWidgets: (amount, refID) {
         
            return const SuccessPaymentWidget();
            
          }),
    );
  }
}

class SuccessPaymentWidget extends StatelessWidget {
  const SuccessPaymentWidget({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        const Icon(
          Icons.check,
          color: Colors.green,
        ),
        const Text('پرداخت شما موفق بود ! '),
        const SizedBox(
          height: 12,
        ),
        const Divider(),
        TextButton(
            onPressed: () {}, child: const Text('بازگشت به صفحه فروشگاه')),
      ],
    );
  }
}

class FailedPaymentWidget extends StatelessWidget {
  const FailedPaymentWidget({
    Key? key,
    required this.themeData,
  }) : super(key: key);

  final ThemeData themeData;

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Icon(
          Icons.error,
          color: themeData.errorColor,
        ),
        Text(
          'پرداخت شما ناموفق بود :(',
          style: themeData.textTheme.subtitle1!.copyWith(
            color: themeData.errorColor,
          ),
        ),
        const Divider(),
        TextButton(onPressed: () {}, child: const Text('بازگشت و تلاش مجدد')),
      ],
    );
  }
}
